typing/typecore.ml

```diff
@@ -1342,7 +1346,7 @@ and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
   | Ppat_or(sp1, sp2) ->
       let state = save_state env in
```

Comment:
This `save_state` was called also on the local or-pattern children. The new implementation calls it only one for the first or-pattern in the local group, and then restores *that* state for each alternative below the group. I suppose that `save_state` is idempotent (calling it twice in a row is the same as saving only once) so this should be correct.

Reply:
Trying to check that `save_state` is indempotent: it's clear for the `env` and `levels` field, but I don't understand the `snapshots` part. Going to assume that it is indeed indempotent (it would be a serious issue if it wasn't).

Reply:
A snapshot is only returning the current contents of the trail pointer, or allocating it if it does not yet exist.
So yes, it is clearly idempotent.
This is standard Prolog implementation technology from the last century :-)